src/main/java/com/salesforce/dynamodbv2/mt/util/RecordsCache.java

```diff
@@ -0,0 +1,276 @@
+        assert shardLocation != null && limit > 0;
+
+        final Lock readLock = lock.readLock();
+        readLock.lock();
+        try {
+            final ShardId shardId = shardLocation.getShardId();
+            final NavigableMap<SequenceNumber, Segment> shardCache = segments.get(shardId);
+            if (shardCache == null) {
+                // nothing cached for the requested shard
+                return Collections.emptyList();
+            }
+
+            final SequenceNumber sequenceNumber = shardLocation.getSequenceNumber();
+            final Entry<SequenceNumber, Segment> entry = shardCache.floorEntry(sequenceNumber);
+            if (entry == null) {
+                // no segment with requested or smaller sequence number
+                return Collections.emptyList();
+            }
+
+            final Segment segment = entry.getValue();
+            if (segment.getEnd().precedes(sequenceNumber)) {
+                // preceding segment does not include requested sequence number
+                return Collections.emptyList();
+            }
+
+            // preceding segment contains (some) records for the requested sequence number
+            final List<Record> records = new ArrayList<>(limit);
+            addAll(records, segment.getRecords(sequenceNumber), limit);
+
+            // keep going through adjacent segments (if present), until limit is reached
+            while (records.size() < limit) {
+                final Segment next = shardCache.get(segment.getEnd().next());
+                if (next == null) {
+                    break;
+                }
+                addAll(records, next.getRecords(), limit);
+            }
+
+            return records;
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    // Should we bring back segment merging to avoid cache fragmentation?
+    void putRecords(ShardLocation shardLocation, List<Record> records) {
+        assert shardLocation != null && records != null && !records.isEmpty();
+
+        final Lock writeLock = lock.writeLock();
+        writeLock.lock();
+        try {
+            final ShardId shardId = shardLocation.getShardId();
+            final SequenceNumber sequenceNumber = shardLocation.getSequenceNumber();
+
+            // lookup segments that immediately precede and succeed new segment to drop overlapping records
+            final Segment previousSegment;
+            final Segment nextSegment;
+            final NavigableMap<SequenceNumber, Segment> shardCache = segments.get(shardId);
+            if (shardCache == null) {
+                previousSegment = null;
+                nextSegment = null;
+            } else {
+                previousSegment = getValue(shardCache.floorEntry(sequenceNumber));
+                nextSegment = getValue(shardCache.higherEntry(sequenceNumber));
+            }
+
+            final Segment segment = new Segment(sequenceNumber, records).subSegment(previousSegment, nextSegment);
```

Comment:
This is another behavior change compared to master: this implementation does not merge segments (see https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_streams_GetRecords.html). It seems less important to merge, given that the read code now iterates over adjacent segments. But we may still want to evaluate/measure the tradeoffs. For example, not merging likely makes writes faster, reducing the time in the critical section. On the other hand, having too many small segments could increase the time it takes to read and evict segments, since more lookups in the sorted map are needed.